src/token_contract/src/main.nr

```diff
@@ -676,4 +744,836 @@ pub contract Token {
+            &mut context,
+        );
+
+        // Mint shares in private
+        _mint_to_private(&mut context, storage.private_balances, from, to, shares);
+
+        // Take the assets from the sender
+        Token::at(storage.asset.read())
+            .transfer_private_to_public(from, context.this_address(), assets, nonce)
+            .call(&mut context);
+    }
+
+    /// @notice Deposits the underlying asset from a private balance and receives shares to a public balance
+    /// @param from The address of the sender
+    /// @param to The address of the recipient
+    /// @param assets The amount of assets to deposit
+    /// @param nonce The nonce used for authwitness
+    #[private]
+    fn deposit_private_to_public(from: AztecAddress, to: AztecAddress, assets: u128, nonce: Field) {
+        // Calculate and mint shares to the recipient in public
+        Token::at(context.this_address()).mint_shares_internal(to, assets).enqueue(&mut context);
+
+        // Take the assets from the sender
+        Token::at(storage.asset.read())
+            .transfer_private_to_public(from, context.this_address(), assets, nonce)
+            .call(&mut context);
+    }
+
+    /// @notice Deposits the underlying asset from a public balance and receives shares to a private balance
+    /// @dev min_shares is the amount of shares that will be minted immediately in private.
+    ///      Any additional shares the recipient is entitled to, will be minted through a commitment once the call is executed publicly.
+    ///      This additional amount can only be known at public execution.
+    /// @param from The address of the sender
+    /// @param to The address of the recipient
+    /// @param assets The amount of assets to deposit
+    /// @param min_shares The amount of shares that should be minted to the recipient in private
+    /// @param nonce The nonce used for authwitness
+    #[private]
+    fn deposit_public_to_private_exact(
+        from: AztecAddress,
+        to: AztecAddress,
+        assets: u128,
+        min_shares: u128,
+        nonce: Field,
+    ) {
+        // Increase recipient's private balance with `min_shares`
+        _increase_private_balance(&mut context, storage.private_balances, from, to, min_shares);
+
+        // TODO (optimization): partial note logs and validity_commitment nullifier are not needed
+        // Initialize commitment that allows to mint outstanding shares to the recipient in public
+        let commitment =
+            _initialize_transfer_commitment(&mut context, storage.private_balances, from, to);
+
+        // Mint any outstanding shares without revealing the recipient and updates the shares total supply
+        // Reverts if min_shares is greater than allowed
+        Token::at(context.this_address())
+            .finalize_mint_with_commitment_internal(commitment, assets, min_shares)
+            .enqueue(&mut context);
```

Comment:
this should be another method, becaese "finalize mint" shouldn't be called with `assets` and `min_shares` (just with `amount`)

Reply:
`mint` is reserved for erc20 naming space imo, it doesn't have a `min_shares` it just receives a number and emits that amount of tokens, here, it is confusing, because you're naming "mint" some method that's more of a calculation, i'd prefer sticking with "issue" or "deposit" depending on what's exact-in / exact-out

it'd be good to explain in a comment / table what's the difference between deposit / withdraw / issue / redeem